tests/test_date.py

```diff
@@ -2,66 +2,138 @@
+        param(begin=datetime(2014, 6, 15), end=datetime(2014, 6, 25), expected_length=10)
+    ])
+    def test_date_range(self, begin, end, expected_length):
+        self.given_period_start(begin)
+        self.given_period_end(end)
+        self.when_date_range_generated()
+        self.then_range_length_is(expected_length)
+        self.then_all_dates_are_present_in_range()
+        self.then_range_is_in_ascending_order()
 
-    def test_should_one_date_for_each_month(self):
-        fixtures = [
-            (datetime(2014, 4, 15), datetime(2014, 6, 25),
-             [(2014, 4), (2014, 5), (2014, 6)]),
+    @parameterized.expand([
+        param(begin=datetime(2014, 4, 15),
+              end=datetime(2014, 6, 25),
+              expected_months=[(2014, 4), (2014, 5), (2014, 6)]),
+        param(begin=datetime(2014, 4, 25),
+              end=datetime(2014, 5, 5),
+              expected_months=[(2014, 4), (2014, 5)]),
+        param(begin=datetime(2014, 4, 5),
+              end=datetime(2014, 4, 25),
+              expected_months=[(2014, 4)]),
+        param(begin=datetime(2014, 4, 25),
+              end=datetime(2014, 6, 5),
+              expected_months=[(2014, 4), (2014, 5), (2014, 6)]),
+    ])
+    def test_one_date_for_each_month(self, begin, end, expected_months):
+        self.given_period_start(begin)
+        self.given_period_end(end)
+        self.given_period_size(months=1)
+        self.when_month_range_generated()
+        self.then_expected_months_are(expected_months)
 
-            (datetime(2014, 4, 25), datetime(2014, 5, 5),
-             [(2014, 4), (2014, 5)]),
+    @parameterized.expand([
+        'year',
+        'month',
+        'week',
+        'day',
+        'hour',
+        'minute',
+        'second',
+    ])
+    def test_should_reject_easily_mistaken_dateutil_arguments(self, invalid_period):
+        self.given_period_start(datetime(2014, 6, 15))
+        self.given_period_end(datetime(2014, 6, 25))
+        self.given_period_size(**{invalid_period: 1})
+        self.when_period_is_parsed()
+        self.then_period_was_rejected(invalid_period)
 
-            (datetime(2014, 4, 5), datetime(2014, 4, 25),
-             [(2014, 4)]),
+    def given_period_start(self, begin):
+        self.period_start = begin
 
-            (datetime(2014, 4, 25), datetime(2014, 6, 5),
-             [(2014, 4), (2014, 5), (2014, 6)]),
-        ]
+    def given_period_end(self, end):
+        self.period_end = end
 
-        for begin, end, expected in fixtures:
-            result = list(date.date_range(begin, end, months=1))
-            self.assertEquals(expected,
-                              [(d.year, d.month) for d in result])
+    def given_period_size(self, **params):
+        self.period_size = params
 
-    def test_should_reject_easily_mistaken_dateutil_arguments(self):
-        begin = datetime(2014, 6, 15)
-        end = datetime(2014, 6, 25)
+    def when_month_range_generated(self):
```

Comment:
`when_*` methods are describing and action and usually this means using a verb. Also, can you describe what "dates" are you talking about here?